node/common/src/service/archive.rs

```diff
@@ -0,0 +1,490 @@
+use mina_p2p_messages::v2::{self, ArchiveTransitionFronntierDiff};
+use node::core::{channels::mpsc, thread};
+use std::net::SocketAddr;
+
+use super::NodeService;
+
+pub struct ArchiveService {
+    archive_sender: mpsc::UnboundedSender<ArchiveTransitionFronntierDiff>,
+}
+
+const ARCHIVE_SEND_RETRIES: u8 = 5;
+const MAX_EVENT_COUNT: u64 = 100;
+const RETRY_INTERVAL_MS: u64 = 1000;
+
+impl ArchiveService {
+    fn new(archive_sender: mpsc::UnboundedSender<ArchiveTransitionFronntierDiff>) -> Self {
+        Self { archive_sender }
+    }
+
+    fn run(
+        mut archive_receiver: mpsc::UnboundedReceiver<ArchiveTransitionFronntierDiff>,
+        address: SocketAddr,
+    ) {
+        while let Some(breadcrumb) = archive_receiver.blocking_recv() {
+            let mut retries = ARCHIVE_SEND_RETRIES;
+            while retries > 0 {
+                match rpc::send_diff(address, v2::ArchiveRpc::SendDiff(breadcrumb.clone())) {
+                    Ok(result) => {
+                        if result.should_retry() {
+                            node::core::warn!(node::core::log::system_time(); summary = "Archive suddenly closed connection, retrying...");
```

Comment:
`node::core::log::system_time()` is not the default if omitted btw, Sebastien added that [here](https://github.com/openmina/openmina/commit/3c97aab7e713d2099b75bde46686d8eb19950b29)